### PR TITLE
graph.Graph build_* constructors

### PR DESCRIPTION
**Pull request by martinfleis**

### CODE REVIEW

**Comment by martinfleis** [comment]
FYI, the result of `_kernel` is currently misaligned to the original data. e.g. in the KNN case, the adjacency is equal but the `_adjacency` table is incorrectly sorted leading to `sparse` being wrong.

**Comment by jGaboardi** [comment]
> FYI, the result of `_kernel` is currently misaligned to the original data. e.g. in the KNN case, the adjacency is equal but the `_adjacency` table is incorrectly sorted leading to `sparse` being wrong.

Should we open a fresh issue for this?

**Comment by martinfleis** [comment]
I fixed that in the meantime. Need to write a test and push.

File: libpysal/graph/tests/test_builders.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+import geopandas as gpd
+import pandas as pd
+import geodatasets
+from libpysal import graph
+
+TRIANGULATIONS = ["delaunay", "gabriel", "relative_neighborhood", "voronoi"]
+
+"""
+This file tests Graph initialisation from various build_* constructors. The correctness
```
[the remark this answers is not in the document]
Comment (martinfleis) [reply libpysal/graph/tests/test_builders.py:11]: Yes. Ensuring that whatever happens in `base.py` is what we expect. Some Graphs are built via arrays, some via sparse so this ensures it all looks the same in the end.

File: libpysal/graph/base.py [hunk 2]
```diff
@@ -425,17 +444,43 @@ def build_triangulation(
...
         Graph
             libpysal.graph.Graph
         """
-        return NotImplementedError
+        if hasattr(data, "index"):
+            ids = data.index
+        else:
+            ids = pd.RangeIndex(0, len(data))
+
+        if method == "delaunay":
```
[the remark this answers is not in the document]
Comment (martinfleis) [reply libpysal/graph/base.py:465]: Yeah! I was actually looking for that to figure out if I can already do it or not.

File: libpysal/graph/tests/test_base.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
 import string
 
 import pandas as pd
+import geopandas as gpd
+import geodatasets
```
[the remark this answers is not in the document]
Comment (martinfleis) [reply libpysal/graph/tests/test_base.py:5]: > As an aside, what's the current relationship between geodatasets and [examples](https://github.com/pysal/libpysal/tree/geographs/libpysal/examples)?

None? @sjsrey explored that but nothing has been done afaik.

File: libpysal/graph/_triangulation.py [hunk 4]
```diff
@@ -2,21 +2,20 @@
...
 from libpysal.cg import voronoi_frames
 
 try:
-    from numba import njit
+    from numba import njit  # noqa E401
 except ModuleNotFoundError:
     from libpysal.common import jit as njit
 
-_VALID_GEOMETRY_TYPES = ("Point")
+_VALID_GEOMETRY_TYPES = ["Point"]
```
[the remark this answers is not in the document]
Comment (jGaboardi) [reply libpysal/graph/_triangulation.py:18]: I'm on board with changing to a `list` in `_kernel.py` then.

**Comment by martinfleis** [comment]
From further testing, it is clear that `_kernel.py` will need more care to ensure efficient queries, filtering of 0s and island handling but not within this PR. Plus that management of collocated points @ljwolf was cooking.

I'll add missing docstrings and this should be ready to go.

**Review by jGaboardi — APPROVED** [review]

**Comment by ljwolf** [comment]
> colocated points

Still cooking, but this would only affect the KNN "branch" (`if k is not None`)

> efficient queries

It would help to understand what this means? This implementation was adopted because it was quite a bit faster than our existing `Kernel()`... but it's still the case that a kernel with no tapering is a full `N x N` matrix. 

On isolates, I notice the `.eliminate_zeros()` now at the end of the class, which is definitely the right thing to do. I think that the `from_sparse()` method, then, needs to handle filling the empty rows' diagonals with a structural zero. This also reminds me that `from_arrays()` needs to have some way of recording/recognizing that the last `m` row/columns of the adjacency matrix are isolates... which might be done through a `shape` argument?

**Comment by martinfleis** [comment]
> > efficient queries

> It would help to understand what this means?

Nothing in the end. I thought we'd expose distance band as a specific case of kernel as discussed, in which case we would need a different query but I have it now implemented on a local branch separately.


> On isolates, I notice the .eliminate_zeros() now at the end of the class, which is definitely the right thing to do. I think that the from_sparse() method, then, needs to handle filling the empty rows' diagonals with a structural zero.

yup, it is likely that isolates are not yet properly handled in all cases.

> This also reminds me that from_arrays() needs to have some way of recording/recognizing that the last m row/columns of the adjacency matrix are isolates... which might be done through a shape argument?

What do you mean "the last m row/columns"? Isolates are not necessarily last in any representation as far as I understand.

**Comment by knaaptime** [comment]
> On isolates, I notice the .eliminate_zeros() now at the end of the class, which is definitely the right thing to do. I think that the from_sparse() method, then, needs to handle filling the empty rows' diagonals with a structural zero. This also reminds me that from_arrays() needs to have some way of recording/recognizing that the last m row/columns of the adjacency matrix are isolates... which might be done through a shape argument?

this is exacty what i was struggling with in the shoehorn-implementation i originally drafted. There, the isolates always get appended to the end of the adjlist, so thats why i ended up needing the [sort](https://github.com/pysal/libpysal/pull/513#discussion_r1204957311) method

**Comment by martinfleis** [comment]
We could potentially use something like I've implemented [here](https://github.com/martinfleis/libpysal/blob/5086103bb038cfbf8a7b291ad2d75bfb9da82ae0/libpysal/graph/_distance.py#L51-L59) instead of `eliminate_zeros()`. That would ensure that isolates are kept in place.

**Comment by knaaptime** [comment]
(we talked about why we dont want to expose sorting, but if we *could*, it would make [this](https://github.com/pysal/esda/pull/259#discussion_r1284582886) dead simple, because we could always re-align the graph with the df after subsetting the graph)

**Comment by ljwolf** [comment]
> nothing in the end

So, distance band should still allow for kernels underneath the banding, right? Basically, we have three possible implementations: 
- kernel with no tapering (full kernel)
- kernel with KNN tapering (KNN weights, possibly with a kernel function of distance)
- kernel with radius tapering (DistanceBand weights, possibly with a kernel function of distance)

It'd be good to figure out a way to keep this API consistent... I had thought that distance banding could be achieved using a kernel weight with a boxcar kernel? 


> what do you mean? 

For the graph `A <-> B <-> C` with an isolate `D`, the adjacency matrix is:

```python
import numpy

A = numpy.array([
    [0,1,0,0]
    [1,0,1,0]
    [0,1,0,0]
    [0,0,0,0]
    ])
```

If we pass arrays:
```python
i,j = A.nonzero()
```
Then we only know about A,B,C, not D. We would need to provide either `ids` or, if there are no ids, `n`? Hence, the sparse matrix constructor takes a `shape=(n,m)` argument to ensure that if `i.max() < n` or `j.max() < m`, the empty end rows/columns are constructed.

**Comment by knaaptime** [comment]
> Hence, the scipy.sparse matrix constructors take a shape=(n,m) argument to ensure that if i.max() < n or j.max() < m, the empty end rows/columns are constructed.

and this is how i ended up doing it. Instantiate a matrix that includes rows of zeros, then tell the W what those mean.

so it feels like if we're going to expose id2i, we might as well include sort... because any time we need sparse, the ordering matters and sometimes its useful to control that?

**Comment by martinfleis** [comment]
> I had thought that distance banding could be achieved using a kernel weight with a boxcar kernel?

it could but you don't know K to pass to `query` so you'd need to get all distances and use `boxcar` to filter that. This is what I meant by more efficient queries above. We should use `KDTree.sparse_distance_matrix` for distance band instead.

 > For the graph A <-> B <-> C with an isolate D

Yeah, that is not a solved case at the moment. All these edge cases will need special attention but I guess it is easier to figure out the solution on top of basic implementation.

> so it feels like if we're going to expose id2i, we might as well include sort... because any time we need sparse, the ordering matters and sometimes its useful to control that?

`id2i` should probably be private imho but I just needed to use it in other methods. I am still not sure about `sort`.

**Comment by ljwolf** [comment]
What about treating knn and distance band as ways to build a "precomputed" distance matrix for kernel to apply a kernel function to?

This would let knn code to live in the knn function, and likewise for radius/distanceband, leaving kernel simpler for all the rest of the methods that pass into it?

**Comment by martinfleis** [comment]
That sounds nice. Let me give it a go since I am touching this as we speak. Will open a PR to discuss on top of code.

**Comment by martinfleis** [comment]
A draft of treating knn and distance band as precomputed matrices is in https://github.com/pysal/libpysal/pull/544.

**Pull request merged**